### ISSUE TITLE
Use of outer type argument in extension inline unapply(Seq) method causes compiler error

### ISSUE THREAD

**Issue by gorilskij** [lampepfl/dotty#15188]
## Compiler version

#15053

## Minimized code

```Scala 
object O

extension [T] (ctx: O.type) inline def unapplySeq(input: T): Option[Seq[T]] = Some(Seq(input)) 

@main
def Main = {
  val O(x) = 3
  println(s"x: $x")
}
```

## Output

Compiler error
```
Exception in thread "main" java.lang.IndexOutOfBoundsException: 0
        at scala.collection.LinearSeqOps.apply(LinearSeq.scala:117)
        at scala.collection.LinearSeqOps.apply$(LinearSeq.scala:114)
        at scala.collection.immutable.List.apply(List.scala:79)
        at dotty.tools.dotc.core.Substituters$.substParams(Substituters.scala:151)
        at dotty.tools.dotc.core.Substituters$SubstParamsMap.apply(Substituters.scala:197)
        at dotty.tools.dotc.core.Substituters$SubstParamsMap.apply(Substituters.scala:197)
        at scala.collection.immutable.List.mapConserve(List.scala:472)
        at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5553)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5582)
        at dotty.tools.dotc.core.Substituters$.substParams(Substituters.scala:161)
        at dotty.tools.dotc.core.Substituters$SubstParamsMap.apply(Substituters.scala:197)
        at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5555)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5582)
        at dotty.tools.dotc.core.Substituters$.substParams(Substituters.scala:161)
        ...
```

## Expectation

```
x: 3
```

**Comment by prolativ** [comment]
The issue does not occur if `unapplySeq` is non-`inline` or if it's defined as an ordinary method rather than an extension

**Comment by prolativ** [comment]
Compiler version: 3.2.0-RC1-bin-20220514-bb29e20-NIGHTLY

**Comment by gorilskij** [comment]
@rmgk
That's a different side of the same issue, see #8577
Both should eventually be resolved by #15191

**Comment by nicolasstucki** [comment]
We should check if #16358 fixes this
